Code adapt.v1.1

galtimur opened the pull request.

galtimur committed:
1. Splitted data module, moved dataloaders to dataloaders.py.
2. Renamed classes
3. Moved some functions for training to a separate module train_functions.py
4. Annotated variables types, checked by mypy. Some errors still exist.

galtimur committed:
1. Splitted data module, moved dataloaders to dataloaders.py.
2. Renamed classes
3. Moved some functions for training to a separate module train_functions.py
4. Annotated variables types, checked by mypy. Some errors still exist.

galtimur committed:
deleted excessive file test_retrieve.py

galtimur committed:
run black and excluded mypy execution in workflows

galtimur committed:
Added configuration file.
Now paths to data are loaded from config.yaml
pyproject.toml ammended by omegaconf library

galtimur committed:
Created file and added "isort ." and "black ." to .git/pre-push

galtimur committed:
ran isort and black

galtimur committed:
Moved some paths in retrieved to config. Added command-line option that sends config filename to a script.

galtimur committed:
1. Added some tests
2. Replace autofaiss library by vanilla faiss
3. Renamed some variables and functions
4. Precalculation of KNNs is now optional

galtimur committed:
Added validation on random chuncks

egor-bogomolov commented on retro_pytorch/train_functions.py:
```diff
@@ -0,0 +1,100 @@
+    return losses_val_cur
+
+
+def val_upadate(
```
rename to `val_update`

egor-bogomolov commented on scripts/train.py:
```diff
@@ -1,40 +1,54 @@
+f_val = open(filename_val, "a")
 
+## TODO I am not sure that at this step I want to move model hyperparameters to a config file - it is easier for me to read it here.
+# instantiate RETRO, fit it into the TrainingWrapper with correct settings
```
Yet you cannot try different configurations without rewriting the file or duplicating code. Config will look exactly like this, and here you would do something like `retro = RETRO(**config.model_parameters)`

egor-bogomolov commented on scripts/train.py:
```diff
@@ -53,162 +67,113 @@
-    chunks_memmap_path=texts_folder + "train.chunks.dat",  # path to chunks
-    seqs_memmap_path=texts_folder + "train.seq.dat",  # path to sequence data
-    doc_ids_memmap_path=texts_folder
+    chunks_memmap_path=paths.texts_folder + "train.chunks.dat",  # path to chunks
```
Consider using os.path.join to join paths, it will save you from accidentally getting paths like "path/to/foldertrain.chunks.dat"

egor-bogomolov commented on scripts/train.py:
```diff
@@ -53,162 +67,113 @@
 # %%
 
+### setting up number of steps. freq_val - frequency of validation, num_val - number of validation steps
+freq_val = 1000
```
It would be nice to have these numbers in the configuration file

egor-bogomolov commented on retro_pytorch/train_functions.py:
```diff
@@ -0,0 +1,100 @@
+    fetch_neighbours: Callable[[torch.Tensor, torch.Tensor], torch.Tensor],
+    aggregate: list[tuple[torch.Tensor, torch.Tensor]],
+) -> list[float]:
+    model.eval()
```
You should also use `torch.no_grad()` or `torch.inference_mode()` to avoid gradient computations. Also, you do backprop from loss even if you call calc_loss in validation, please avoid it.

egor-bogomolov committed:
Use "heads" parameter in RETRO

egor-bogomolov commented on scripts/train.py:
```diff
@@ -53,162 +67,113 @@
 optim, scheduler = wrapper_db.get_optimizer(warmup_steps=warmup_steps, training_steps=total_steps, lr=lr, wd=0.01)
 scheduler.step()
 fetch_neighbours = wrapper_db.fetch_neighbours
+fetch_random_chunk = wrapper_db.fetch_random_chunk
+generate_pure_random_chunk = wrapper_db.fetch_random_chunk
```
I guess these two should be different

egor-bogomolov reviewed (changes requested):
Left a few comments, please fix training on the validation data and usage of random retrieval functions 🙂

galtimur committed:
Corrected according to comments
1. Corrected the error with validation.
2. Moved almost all numerical (hyper)parameters to config file.
3. Used os.path.join to join pathes
4. Corrected some typos

galtimur committed:
1. Freezed the encoder parameters for no-retrieve regime.2. Fixed some leftovers from previous commit

galtimur committed:
fixed one typo

galtimur committed:
fixed some typo and styling

galtimur committed:
Switched off retrieving random on validation in no-retrieve mode
